GPII-3354 Restrict permissions to the common SA

amatas opened the pull request:
The common SA has owner permissions at organization level and also at the project level.
At the organization level it needs the `owner` permissions in order to grant itself the permissions of modifying IAMs. Once the common SA has the right permissions it doesn't need to reset them again, so we can delegate this task to an operator and remove the owner permission to the SA at organization level. If the common SA doesn't have proper permissions the CI pipeline will fail at the `apply_common_infra` task.

At the project level it also has the `owner` role. This PR set the roles collected based on the audit logs queries and some test deployments. This PR also assign lower level permissions to the user SA.

The plan to proceed with this PR, as it affects to the entire infrastructure, is to merge and perform all the checks in the common-STG environment before it its promoted to production. The checks can be:

- [ ] activate audit logs for the doe dev cluster and the test organization
- [ ] rake command in the doe dev cluster
- [ ] check that the endpoints work
  - [ ] rake test_flowmanager
  - [ ] rake test_prefences
- [ ] rake destroy in the doe dev cluster
- [ ] remove all the resources left in the doe cluster manually. The project must be as cleaned as possible of resources, like a brand new project.
- [ ] Test again the rake and rake destroy commands
- [ ] See that the logs are sent to the Stackdriver
- [ ] Verify that the logs have been exported to the Google Storage Bucket
- [ ] check that all the permissions were granted at the project level an at the organization level.

Perhaps this PR also covers part of GPII-2802

mrtyler commented:
>  check that the endpoints work

I would make this step more specific. I would suggest running `rake test_flowmanager` and `rake test_prefences`, plus whatever other manual endpoint checks are required.

> remove all the resources left in the doe cluster manually

Can you elaborate? I think this means "do what `rake destroy_infra` would do if it worked"?

> Test again the rake and rake destroy commands

Is this to verify that `rake` works from scratch, i.e. when running `rake apply_infra` for the first time?

> See that the logs are sent to the Stackdriver

Can you also please verify that exports are exporting?

> Perhaps this PR also covers part of GPII-2802

I agree. I added a link from that ticket to this PR.

mrtyler commented on shared/rakefiles/xk_infra.rake:
```diff
@@ -84,11 +84,37 @@ task :apply_common_infra => [@gcp_creds_file] do
+
+  organizations_permissions = ["roles/resourcemanager.projectCreator", "roles/billing.user"]
+
+  serviceAccount = "serviceAccount:projectowner@#{ENV["TF_VAR_project_id"]}.iam.gserviceaccount.com"
```
Variable names should be snake_case, `service_account`.

mrtyler commented on shared/rakefiles/xk_infra.rake:
```diff
@@ -84,11 +84,37 @@ task :apply_common_infra => [@gcp_creds_file] do
+    organizations_permissions.delete(permission["role"]) if organizations_permissions.include?(permission["role"]) and permission["members"].include?(serviceAccount)
   end
 
+  raise "The common Service Account #{serviceAccount} doesn't have the proper permissions #{organizations_permissions}" unless organizations_permissions.empty?
```
A good error message offers suggestions on how to correct the error.

Can you add something like "An admin must run the following:" and include the now-commented-out command from L92-94? You'll need to change the command to a runnable form, `rake sh"[for role in ... ; do gcloud organizations add-iam-policy-binding ...]"`.

Then, you can consolidate L87-100 to explain why the common SA needs permissions that it cannot grant itself.

These changes help the user seeing the error message while giving us some chance of keeping the manual step up-to-date (something less likely to happen if it lives as a block of commented-out code).

stepanstipl replied:
or even include rake task to do the one-time setup?

mrtyler commented on shared/rakefiles/xk_infra.rake:
```diff
@@ -84,11 +84,37 @@ task :apply_common_infra => [@gcp_creds_file] do
+  }
+
+  permissions_list = JSON.parse(permissions_list_json)["bindings"]
+
```
I would remove all these blank lines. L106-116 are one logical block.

amatas commented:
> > check that the endpoints work
>
> I would make this step more specific. I would suggest running `rake test_flowmanager` and `rake test_prefences`, plus whatever other manual endpoint checks are required.
>
> > remove all the resources left in the doe cluster manually
>
> Can you elaborate? I think this means "do what `rake destroy_infra` would do if it worked"?
> 
> > Test again the rake and rake destroy commands
> 
> Is this to verify that `rake` works from scratch, i.e. when running `rake apply_infra` for the first time?
> 
> > See that the logs are sent to the Stackdriver
> 
> Can you also please verify that exports are exporting?


I changed the description of the PR.
I also created a rake task for setting up the common SA permissions and added some more comments in the code and in the exception to recover from a possible failure related to the wrong permissions settings.

mrtyler commented on shared/rakefiles/xk_infra.rake:
```diff
@@ -98,22 +98,16 @@ task :apply_common_infra => [@gcp_creds_file] do
   end
-
-  raise "The common Service Account #{serviceAccount} doesn't have the proper permissions #{organizations_permissions}" unless organizations_permissions.empty?
+  raise "The common Service Account #{service_account} doesn't have the proper permissions #{organizations_permissions}.\n\nAn operator with admin privileges on the organization must run the following command: rake fix_common_service_account_permissions\n\nCheck the credentials used with the command: rake sh[\"gcloud auth list\"]\n\nThe common Service Account needs these permissions at organizati ...
```
This is a good error message. :)

I think the last sentence is clear from context, so I would delete it.

I'm not sure exactly what this is telling me: `Check the credentials used with the command: rake sh[\"gcloud auth list\"]`. I like that the command is included, but what am I looking for in the output? (I think what you're after is that admins must be authenticated as themselves, not as a Service Account with insufficient permissions. I'm not sure that will be clear to everyone.)

While you're editing, can you also break this up into some shorter lines?

mrtyler commented on shared/rakefiles/xk_infra.rake:
```diff
@@ -98,22 +98,16 @@ task :apply_common_infra => [@gcp_creds_file] do
   # permssions to an operator. The following code checks that the SA has the
```
Most of this comment (L87-100) is now documented elsewhere. Can you edit it down or delete it?

mrtyler commented:
Oh, and the updates to the deployment plan look good. Thanks.

stepanstipl commented:
> Is there a role that covers service account creation? (like roles/iam.serviceAccountAdmin)

@amatas did you see my comment? I'm just worried we're missing some permissions

amatas commented:
yes @stepanstipl I saw your comment, but I think that the common SA doesn't need any more permissions.

mrtyler commented:
LGTM

natarajaya commented:
LGTM

amatas commented:
My last change uses the most authoritative resource to set the IAMs, because the google_project_iam_binding for both SA was causing some fighting between the two resources to set the proper permissions. Having all the permissions in only one block avoid conflicts and also it's applied faster.

I have been searching how to interpolate that resource in order to make it more readable but I could find anything useful.

I also tested the code in the testing organization creating a dev project called "test1", I've also tested a complete deployment and destroying. So it seems that the permissions are properly set.

amatas commented:
@natarajaya @stepanstipl :point_up: plz

stepanstipl commented:
yep, looks good

natarajaya commented on shared/rakefiles/xk_infra.rake:
```diff
@@ -97,3 +107,11 @@ task :apply_common_infra => [@gcp_creds_file] do
   end
 end
+
+task :fix_common_service_account_permissions => [@gcp_creds_file] do
```
I don't really like having two tasks with the same name in different rake files, feels a little bit confusing. But since this is a side effect of having part of the code that runs inside of the xk container, I don't have a better proposition at the moment...

natarajaya commented:
> google_project_iam_binding for both SA was causing some fighting between the two resources to set the proper permissions. Having all the permissions in only one block avoid conflicts and also it's applied faster.

Nice catch!

LGTM
